case-lib/lib.sh

```diff
@@ -682,26 +682,10 @@ is_zephyr()
 
 is_firmware_file_zephyr()
 {
-    local firmware_name firmware_path znum platform
-    # get the FW name and path
-    # TODO: optimize this part after driver can expose the FW path to userspace via debugfs
-    if is_ipc4; then
-        firmware_name=dsp_basefw.bin
-        fw_mod_para=/sys/module/snd_sof_pci/parameters/fw_path
-        if [ -s "$fw_mod_para" ]; then
-            firmware_path=$(cat $fw_mod_para)
-        else
-            # # FIXME: the kernel driver should give us the FW path
-            # https://github.com/thesofproject/linux/issues/3867
-            die "Failed to get the IPC4 FW path."
-        fi
-    else # for IPC3
-        platform=$(sof-dump-status.py -p)
-        firmware_name=sof-$platform.ri
-        firmware_path=$(sof-dump-status.py -P)
-    fi
+    local firmware_path znum
 
-    znum=$(strings "/lib/firmware/$firmware_path/$firmware_name" | grep -c -i zephyr)
+    firmware_path=$(journalctl_cmd | grep "request_firmware" | awk -F ' ' '{print $(NF-1)}')
+    znum=$(strings "/lib/firmware/$firmware_path" | grep -c -i zephyr)
```

Comment:
what makes you think 'zephyr' is used in the kernel @aiChaoSONG ?

Reply:
the function `is_firmware_file_zephyr()` is used to check if the firmware uses zephyr, if yes，mtrace-reader.py is used for firmware log. btw, I only simplify the function, don't change its behavior.

Reply:
the firmware path is not by default using the 'zephyr' prefix in any kernel descriptor.

Maybe that's what you use in your CI setup, but that would mean that you override the default locations.

That cannot be a solution.

Reply:
@plbossart `strings` is searching the file content, not its path.

Reply:
When we previously align FW and topology layout, we agreed on CI DUTs:
- For SOF Zephyr+ IPC3, driver will request /lib/firmware/intel/sof-ipc3-zephyr/<key>/sof-<platform>.ri
- For SOF Zephyr + IPC4 on cavs platforms, driver will request /lib/firmware/intel/sof-ipc4/<platform name>/[key]/dsp_basefw.bin
- For SOF Zephyr + IPC4 on MTL, driver will request /lib/firmware/intel/sof-ipc4/<platform name>/[key]/sof-<platform name>.ri
- For IPC4 tester, driver will request for /lib/firmware/intel/avs/<platform name>/dsp_basefw.bin

Can we assume developers' personal test device also follow this agreement?  @plbossart @marc-hb @aiChaoSONG 
If yes, the firmware path contains `sof-ipc3-zephyr` or `sof-ipc4`, we know is Zephyr builds. If developers prefer to managing firmware in another way, we cannot totally depend on the firmware path.

Reply:
> @plbossart `strings` is searching the file content, not its path.

The patch looks reasonable. And thanks for updating the commit message @aiChaoSONG

Reply:
This does make a dependency to dev_dbg print that is not on by default. Sof-test won't work without this.

Reply:
@aiChaoSONG It would be good to fail with a clear error if the "request_firmware" line is not found (e.g. dynamic debug not enabled in kernel and/or the line from SOF loader.c is disabled). Now the test will silently conclude it is not Zephyr and do the wrong thing.

Reply:
> It would be good to fail with a clear error if the "request_firmware" line is not found

Indeed.

If the user expects logging and logging fails then the test should fail. Corner cases that want to test without logging must use something like `SOF_LOGGING=none`


`[ -n "$firmware_path" ] || die 'no firmware loaded'`